torch/_functorch/aot_autograd.py

```diff
@@ -3205,7 +3205,7 @@ def runtime_wrapper(*args):
             for idx in indices_of_inps_to_detach:
                 if isinstance(args_[idx], torch.Tensor):
                     args_[idx] = args_[idx].detach()
-            with torch.autograd._force_original_view_tracking(True):
+            with torch.enable_grad(), torch.autograd._force_original_view_tracking(True):
```

Comment:
I don't completely understand why this seems to fix things, but it does.

Reply:
I've added even more changes lessening the dependence on ambient `no_grad` and emphasizing more whether any output `requires_grad`. `needs_autograd` is controlled by the `AOTConfig.allow_joint_graph` which is by default `True`. 

In the export setting, one is required to be explicit about whether the joint graph is traced, as there are restrictions on this joint graph.

In the compile setting, we will create backwards function as long as there is an output with `requires_grad`

Reply:
yep, the `enable_grad` is definitely more correct - if we knew at trace-time that we needed a training graph, then we definitely need the ambient grad state to be turned on at runtime.

If it turns out that some of our outputs should rquire grad and some should not, that should be handled by the fact we're using `ctx.mark_non_differentiable()` to set some outputs as not requiring grad (based on what we detected at tracetime), [here](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py#L507)